src/host/posix/dnssd.cpp

```diff
@@ -672,21 +608,144 @@ void DnssdPlatform::StopAddressResolver(const AddressResolver
     {
         auto &entryList = iter->second;
 
-        entryList.MarkAsDeleted(aAddressResolver.mInfraIfIndex, aCallback);
+        entryList.Delete(aAddressResolver.mInfraIfIndex, aCallback);
+
+        if (entryList.IsEmpty())
+        {
+            aResolversMap.erase(iter);
+        }
+
+        PostHostSubcriptionUpdateTask();
+    }
+}
+
+void DnssdPlatform::ExecuteServiceSubscriptionUpdate(void)
+{
+    mServiceSubscriptionUpdateTaskPosted = false;
+
+    // Unsubscribe services (DnsServiceType) that are stale.
+    for (auto iter = mServiceTypeSubscriptions.begin(); iter != mServiceTypeSubscriptions.end();)
+    {
+        if (mServiceBrowsersMap.find(*iter) == mServiceBrowsersMap.end())
+        {
+            mPublisher.UnsubscribeService(iter->ToString(), "");
+            iter = mServiceTypeSubscriptions.erase(iter);
+        }
+        else
+        {
+            iter++;
+        }
+    }
+
+    // Subscribe to services (DnsServiceType) that haven't been subscribed.
+    for (const auto &entry : mServiceBrowsersMap)
+    {
+        const DnsServiceType &serviceType = entry.first;
 
-        if (!entryList.HasAnyValidCallbacks())
+        if (mServiceTypeSubscriptions.find(serviceType) == mServiceTypeSubscriptions.end())
         {
-            mPublisher.UnsubscribeHost(aAddressResolver.mHostName);
+            mServiceTypeSubscriptions.insert(serviceType);
+            mPublisher.SubscribeService(serviceType.ToString(), "");
         }
+    }
 
-        if (!mInvokingCallbacks)
+    // Unsubscribe services (DnsServiceName) that are stale.
+    for (auto iter = mServiceNameSubscriptions.begin(); iter != mServiceNameSubscriptions.end();)
+    {
+        if (mServiceResolversMap.find(*iter) == mServiceResolversMap.end() &&
+            mTxtResolversMap.find(*iter) == mTxtResolversMap.end())
         {
-            entryList.CleanUpDeletedEntries();
-            if (entryList.IsEmpty())
-            {
-                aResolversMap.erase(iter);
-            }
+            mPublisher.UnsubscribeService(iter->GetType(), iter->GetInstance());
+            iter = mServiceNameSubscriptions.erase(iter);
         }
+        else
+        {
+            iter++;
+        }
+    }
+
+    // Subscribe to services (DnsServiceName) that haven't been subscribed.
+    for (const auto &entry : mServiceResolversMap)
+    {
+        const DnsServiceName &serviceName = entry.first;
+
+        if (mServiceNameSubscriptions.find(serviceName) == mServiceNameSubscriptions.end())
+        {
+            mServiceNameSubscriptions.insert(serviceName);
+            mPublisher.SubscribeService(serviceName.GetType(), serviceName.GetInstance());
```

Comment:
I have an impression that the `mdns::Publisher` doesn't handle the case well when subscribing to the same thing multiple times. Do we have any dependency on this?